fix(agent): Correctly truncate the disk buffer

nagaflokhu opened the pull request:
## Summary

The disk buffer grows and never shrinks, as documented [here](https://github.com/influxdata/telegraf/issues/16696).

## Checklist

- [x] No AI generated code was used in this PR

## Related issues

resolves #16314
resolves #16500
resolves #16615
resolves #16696

nagaflokhu commented:
This fix appears to work in the tests I have run, but I am not confident that it actually always works.

nagaflokhu commented on models/buffer_disk.go:
```diff
@@ -223,7 +221,7 @@ func (b *DiskBuffer) EndTransaction(tx *Transaction) {
 	// Remove the metrics that are marked for removal from the front of the
 	// WAL file. All other metrics must be kept.
-	if len(b.mask) == 0 || b.mask[0] != 0 {
+	if len(b.mask) == 0 {
```
I am not sure why, but if I left the second condition, my fix did not work, despite the changes above and below.

nagaflokhu commented:
the buffer fullness debug messages appear not to work with my patch. e.g.:
2025-03-26T19:34:07Z D! [outputs.file] Buffer fullness: -818 metrics

srebhan commented on models/buffer_disk.go:
```diff
@@ -147,13 +147,11 @@ func (b *DiskBuffer) BeginTransaction(batchSize int) *Transaction {
-	for batchSize > 0 && readIndex < endIndex {
+	for ; batchSize > 0 && readIndex < endIndex; readIndex, offset = readIndex+1, offset+1 {
 		data, err := b.file.Read(readIndex)
 		if err != nil {
 			panic(err)
 		}
-		readIndex++
-		offset++
 
```
This is no functional change. Can we please get this reverted to see the minimal change necessary to fix the problem?

nagaflokhu replied:
you are perhaps right in the case of `readIndex`, but I am pretty sure moving the increment of `offset` is in fact a functional change.

`offset` is being added into a slice in the body of the loops. putting `offset++` earlier in that same loop body guarantees that `offset` can *never be equal to 0* when it is added to the slice, whereas placing `offset++` *outside the body of the loop* means `offset` can be equal to 0 when it gets added to the slice, which seems to be the intended behaviour.

nagaflokhu commented:
it might be a good idea to consider different approaches to dealing with the fact that the `wal` lib does not allow truncating all of its elements. four that come to mind (the first two being very similar to each other):

1. every time there is a write to the WAL, add a "sentinel" element in addition to the metrics being added. the sentinel is a well-known value that is impossible to obtain from serializing a metric. when reading from the WAL, ignore the sentinels. when truncating all metrics from the WAL, make sure to leave the latest write's sentinel in the WAL.
2. when trunctaing all metrics from the WAL, add a sentinel first, and then truncate everything except the sentinel. (why shouldn't this be possible?)
3. change the `wal` library such as to allow truncating everything from the WAL.
4. don't use the WAL library. it seems to me that, at an abstract level, what we want for this buffer is an on-disk deque, i.e. the possibility to add to/read from/remove from the front/back.

srebhan commented on models/buffer_disk.go:
```diff
@@ -147,13 +147,11 @@ func (b *DiskBuffer) BeginTransaction(batchSize int) *Transaction {
 	readIndex := b.batchFirst
 	endIndex := b.writeIndex()
 	offset := 0
-	for batchSize > 0 && readIndex < endIndex {
+	for ; batchSize > 0 && readIndex < endIndex; readIndex, offset = readIndex+1, offset+1 {
```
Oh I see because we are adding this _before_ the rest of the code and you do it at the end... In this case it would be good to keep the `readIndex` as is and use the following

```golang
	for offset :=0; batchSize > 0 && readIndex < endIndex; offset++ {
```

nagaflokhu replied:
done

nagaflokhu replied:
I was wondering if we were going to attempt to complete this PR? it seems like a pretty major issue with the disk buffer

nagaflokhu commented:
!signed-cla

srebhan commented:
@nagaflokhu sorry for this being so slow! We do appreciate your effort on this! For moving forward with the PR, we need to check why the tests are failing now. If the tests were wrong initially, we need an explanation on why they were wrong and a fix to correct the test. Otherwise, if the tests are correct, we need to fix the code to get the tests to pass.
We also need to make sure that the present error is covered by a unit-test, i.e. add a test if there is none yet.

nagaflokhu commented:
@srebhan I've now fixed the tests that were failing and added an assertion for regression purposes. I just ran the additional assertion on the code from the latest master branch commit, and it shows that the previous implementation was not working: accepting a transaction and ending it did not change the amount of entries in the buffer. the other assertion in the test I modified was passing because the mask had a length of 2.

srebhan commented:
@nagaflokhu I took the liberty to push some commits to your branch including a unit-test and some missing pieces for fixing the issue. What do you think?

nagaflokhu commented:
> @nagaflokhu I took the liberty to push some commits to your branch including a unit-test and some missing pieces for fixing the issue. What do you think?

very nice!

fwiw, I wonder if an alternative disk buffer design might not be simpler than having to manage a mask, but it looks to me like we're not going down that route, so let's move forward I guess.

srebhan commented:
An alternative WAL implementation will not happen anytime soon so let's fix this!

srebhan reviewed (approved):
Thanks @nagaflokhu for pushing this forward!

srebhan commented:
@mstrandboge and @skartikey approved this but as I'm involved in the code, please review both! Furthermore, review but do not merge yet until we get some confirmations it fixes the issue(s).

skartikey commented:
The proposed fix looks solid and appears to resolve the issue with the buffer not shrinking properly. It correctly maintains the mask, handles the empty buffer case more consistently, and adjusts offsets after truncation as expected.

While reading through the code, I considered checking whether the condition `b.isEmpty = b.entries()-removeIdx <= 0` reliably detects when the buffer becomes empty. This would be a key area to investigate if shrinking were to fail. However, I later noticed that the `TestDiskBufferTruncate` test case already covers this scenario effectively. It verifies the buffer’s behavior across multiple batches and confirms that, when logically empty, no metrics are returned and the remaining masked entry is handled appropriately due to the WAL design.

The fix aligns well with what this test validates. Great work on this!

srebhan commented:
Thanks @skartikey for the thorough review!
